File: Packages/LocumTrackerCore/Sources/LocumTrackerCore/Models/Location.swift
// LocumTracker
// Copyright (C) 2025 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import Foundation
import SwiftData

/// Modified Monash Model classification levels
public enum MMMClassification: Int, CaseIterable, Codable {
    case majorCity = 1
    case regionalCity = 2
    case largeRuralTown = 3
    case mediumRuralTown = 4
    case smallRuralTown = 5
    case remoteCommunity = 6
    case veryRemoteCommunity = 7

    public var description: String {
        switch self {
        case .majorCity: return "MMM1 - Major City"
        case .regionalCity: return "MMM2 - Regional City"
        case .largeRuralTown: return "MMM3 - Large Rural Town"
        case .mediumRuralTown: return "MMM4 - Medium Rural Town"
        case .smallRuralTown: return "MMM5 - Small Rural Town"
        case .remoteCommunity: return "MMM6 - Remote Community"
        case .veryRemoteCommunity: return "MMM7 - Very Remote Community"
        }
    }

    /// Whether this classification is eligible for rural subsidy (MMM3-7)
    public var isSubsidyEligible: Bool {
        rawValue >= 3
    }
}

/// Represents a physical location where locum work is performed
@Model
public final class Location {
    public var id: UUID = UUID()
    public var name: String = ""
    public var address: String = ""
    public var mmmClassification: Int = 1
    public var latitude: Double?
    public var longitude: Double?
    public var effectiveFrom: Date = Date()
    public var effectiveTo: Date?
    public var createdAt: Date = Date()
    public var updatedAt: Date = Date()

    // MARK: - Contact & Provider Info

    /// Medicare provider number for this location (location-specific)
    public var providerNumber: String?
    /// Phone number for the location
    public var phoneNumber: String?
    /// Free-text notes about this location
    public var notes: String?

    // MARK: - Default Rates

    /// Default daily rate for assignments at this location
    public var defaultDailyRate: Double?
    /// Default hourly rate for assignments at this location
    public var defaultHourlyRate: Double?
    /// Default on-call rate for assignments at this location
    public var defaultOnCallRate: Double?
    /// Default call-out rate for assignments at this location
    public var defaultCallOutRate: Double?

    // MARK: - Default Session Templates

    /// JSON-encoded default session templates (stored as Data for SwiftData compatibility)
    public var defaultSessionTemplatesJSON: Data?

    /// Default session templates for this location.
    /// These define typical work sessions that can be pre-populated when creating daily records.
    public var defaultSessionTemplates: [DefaultSessionTemplate] {
        get {
            guard let data = defaultSessionTemplatesJSON else { return [] }
            return (try? JSONDecoder().decode([DefaultSessionTemplate].self, from: data)) ?? []
        }
        set {
            defaultSessionTemplatesJSON = newValue.isEmpty ? nil : try? JSONEncoder().encode(newValue)
        }
    }

    public init(
        id: UUID = UUID(),
        name: String,
        address: String,
        mmmClassification: Int,
        latitude: Double? = nil,
        longitude: Double? = nil,
        effectiveFrom: Date = Date(),
        providerNumber: String? = nil,
        phoneNumber: String? = nil,
        notes: String? = nil,
        defaultDailyRate: Double? = nil,
        defaultHourlyRate: Double? = nil,
        defaultOnCallRate: Double? = nil,
        defaultCallOutRate: Double? = nil,
        defaultSessionTemplates: [DefaultSessionTemplate] = []
    ) {
        self.id = id
        self.name = name
        self.address = address
        self.mmmClassification = mmmClassification
        self.latitude = latitude
        self.longitude = longitude
        self.effectiveFrom = effectiveFrom
        self.effectiveTo = nil
        self.createdAt = Date()
        self.updatedAt = Date()
        self.providerNumber = providerNumber
        self.phoneNumber = phoneNumber
        self.notes = notes
        self.defaultDailyRate = defaultDailyRate
        self.defaultHourlyRate = defaultHourlyRate
        self.defaultOnCallRate = defaultOnCallRate
        self.defaultCallOutRate = defaultCallOutRate
        if !defaultSessionTemplates.isEmpty {
            self.defaultSessionTemplatesJSON = try? JSONEncoder().encode(defaultSessionTemplates)
        }
    }

    /// Determines if this location is eligible for rural subsidy (MMM 3-7)
    public var isRuralSubsidyEligible: Bool {
        (3...7).contains(mmmClassification)
    }

    /// Returns the MMM classification as a descriptive string
    public var mmmClassificationDescription: String {
        MMMClassification(rawValue: mmmClassification)?.description ?? "Unknown"
    }

    /// Whether this location has any default rates configured
    public var hasDefaultRates: Bool {
        defaultDailyRate != nil || defaultHourlyRate != nil || defaultOnCallRate != nil || defaultCallOutRate != nil
    }

    /// Whether this location has default session templates configured
    public var hasDefaultSessionTemplates: Bool {
        !defaultSessionTemplates.isEmpty
    }
}

File: Android/core/src/commonMain/kotlin/com/hherb/locumtracker/core/model/Location.kt
package com.hherb.locumtracker.core.model

import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable
import com.hherb.locumtracker.core.util.randomUuidString

/** Lowest MMM classification (inclusive) that is eligible for rural subsidies. */
private const val MIN_RURAL_SUBSIDY_MMM = 3

/** Highest MMM classification (inclusive) that is eligible for rural subsidies. */
private const val MAX_RURAL_SUBSIDY_MMM = 7

/** Number of minutes in one hour, used to convert wall-clock times to durations. */
private const val MINUTES_PER_HOUR = 60

/** Number of minutes in one hour as a [Double], used for fractional-hour conversion. */
private const val MINUTES_PER_HOUR_DOUBLE = 60.0

/**
 * A practice location where locum work is performed.
 *
 * The [mmmClassification] (Modified Monash Model, 1-7) drives rural-subsidy
 * eligibility: classifications 3-7 are rural/remote and eligible.
 *
 * @property id Unique identifier for the location.
 * @property name Display name of the location.
 * @property address Postal address of the location.
 * @property mmmClassification Modified Monash Model classification (1-7).
 * @property latitude Optional geographic latitude.
 * @property longitude Optional geographic longitude.
 * @property effectiveFrom Date from which this location record is effective.
 * @property effectiveTo Optional date after which this record is no longer effective.
 * @property createdAt Timestamp when the record was created.
 * @property updatedAt Timestamp when the record was last updated.
 * @property providerNumber Optional Medicare provider number for this location.
 * @property phoneNumber Optional contact phone number.
 * @property notes Optional free-text notes.
 * @property defaultDailyRate Optional default daily rate for assignments at this location.
 * @property defaultHourlyRate Optional default hourly rate for assignments at this location.
 * @property defaultOnCallRate Optional default on-call rate for assignments at this location.
 * @property defaultCallOutRate Optional default call-out rate for assignments at this location.
 * @property defaultSessionTemplatesJSON Serialized default session templates as JSON.
 */
@Serializable
data class Location(
    val id: String = randomUuidString(),
    val name: String = "",
    val address: String = "",
    val mmmClassification: Int = 1,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val effectiveFrom: Instant = Instant.fromEpochMilliseconds(System.currentTimeMillis()),
    val effectiveTo: Instant? = null,
    val createdAt: Instant = Instant.fromEpochMilliseconds(System.currentTimeMillis()),
    val updatedAt: Instant = Instant.fromEpochMilliseconds(System.currentTimeMillis()),
    val providerNumber: String? = null,
    val phoneNumber: String? = null,
    val notes: String? = null,
    val defaultDailyRate: Double? = null,
    val defaultHourlyRate: Double? = null,
    val defaultOnCallRate: Double? = null,
    val defaultCallOutRate: Double? = null,
    val defaultSessionTemplatesJSON: String? = null
) {
    /** `true` when the [mmmClassification] falls within the rural-subsidy-eligible range (MMM3-MMM7). */
    val isRuralSubsidyEligible: Boolean
        get() = mmmClassification in MIN_RURAL_SUBSIDY_MMM..MAX_RURAL_SUBSIDY_MMM

    /** `true` when at least one default rate has been configured for this location. */
    val hasDefaultRates: Boolean
        get() = defaultDailyRate != null || defaultHourlyRate != null ||
                defaultOnCallRate != null || defaultCallOutRate != null

    /** `true` when default session templates have been configured for this location. */
    val hasDefaultSessionTemplates: Boolean
        get() = !defaultSessionTemplatesJSON.isNullOrEmpty()

    /** Human-readable description of the [mmmClassification], or "Unknown" if out of range. */
    val mmmClassificationDescription: String
        get() = when (mmmClassification) {
            1 -> "MMM1 - Major City"
            2 -> "MMM2 - Regional City"
            3 -> "MMM3 - Large Rural Town"
            4 -> "MMM4 - Medium Rural Town"
            5 -> "MMM5 - Small Rural Town"
            6 -> "MMM6 - Remote Community"
            7 -> "MMM7 - Very Remote Community"
            else -> "Unknown"
        }
}

/**
 * A reusable template describing a session's start and end time of day.
 *
 * @property id Unique identifier for the template.
 * @property startHour Hour of day the session starts (0-23).
 * @property startMinute Minute of the hour the session starts (0-59).
 * @property endHour Hour of day the session ends (0-23).
 * @property endMinute Minute of the hour the session ends (0-59).
 * @property label Optional human-readable label for the template.
 */
@Serializable
data class DefaultSessionTemplate(
    val id: String = randomUuidString(),
    val startHour: Int,
    val startMinute: Int,
    val endHour: Int,
    val endMinute: Int,
    val label: String? = null
) {
    /** Duration of the template in fractional hours, or `0.0` if the end is not after the start. */
    val durationHours: Double
        get() {
            val startMinutes = startHour * MINUTES_PER_HOUR + startMinute
            val endMinutes = endHour * MINUTES_PER_HOUR + endMinute
            val durationMinutes = endMinutes - startMinutes
            return if (durationMinutes > 0) durationMinutes.toDouble() / MINUTES_PER_HOUR_DOUBLE else 0.0
        }

    /** The start and end times formatted as "HH:mm - HH:mm". */
    val timeRangeFormatted: String
        get() {
            val startStr = "%02d:%02d".format(startHour, startMinute)
            val endStr = "%02d:%02d".format(endHour, endMinute)
            return "$startStr - $endStr"
        }
}
